File: oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/java/authentication/RedirectUriUtil.java
package ${package}.authentication;

import se.curity.identityserver.sdk.errors.ErrorCode;
import se.curity.identityserver.sdk.service.ExceptionFactory;
import se.curity.identityserver.sdk.service.authentication.AuthenticatorInformationProvider;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import static ${package}.descriptor.${pluginName}AuthenticatorPluginDescriptor.CALLBACK;

final class RedirectUriUtil
{
    private RedirectUriUtil() { }

    static String createRedirectUri(AuthenticatorInformationProvider authenticatorInformationProvider,
                                    ExceptionFactory exceptionFactory)
    {
        try
        {
            URI authUri = authenticatorInformationProvider.getFullyQualifiedAuthenticationUri();

            return new URI(
                    authUri.getScheme(),
                    authUri.getAuthority(),
                    authUri.getPath() + "/" + CALLBACK,
                    authUri.getQuery(),
                    authUri.getFragment()
            ).toURL().toString();
        }
        catch (URISyntaxException | MalformedURLException e)
        {
            throw exceptionFactory.internalServerException(ErrorCode.INVALID_REDIRECT_URI,
                    "Could not create redirect URI");
        }
    }
}

File: kotlin-oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/kotlin/authentication/RedirectUriUtil.kt
package ${package}.authentication

import se.curity.identityserver.sdk.errors.ErrorCode.INVALID_REDIRECT_URI
import se.curity.identityserver.sdk.service.ExceptionFactory
import se.curity.identityserver.sdk.service.authentication.AuthenticatorInformationProvider

import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException

import ${package}.descriptor.${pluginName}AuthenticatorPluginDescriptor.Companion.CALLBACK
class RedirectUriUtil
{
    companion object {
        fun createRedirectUri(authenticatorInformationProvider: AuthenticatorInformationProvider,
                              exceptionFactory: ExceptionFactory): String
        {
            try
            {
                val authUri = authenticatorInformationProvider.fullyQualifiedAuthenticationUri

                return URI(authUri.scheme, authUri.authority, authUri.path + "/" + CALLBACK, authUri.query, authUri.fragment).toURL().toString()
            }
            catch (e: MalformedURLException)
            {
                throw exceptionFactory.internalServerException(INVALID_REDIRECT_URI,
                    "Could not create redirect URI")
            }
            catch (e: URISyntaxException)
            {
                throw exceptionFactory.internalServerException(INVALID_REDIRECT_URI,
                    "Could not create redirect URI")

            }
        }
    }
}
